tensorflow/core/kernels/mkl_matmul_op_fused.cc

```diff
@@ -0,0 +1,194 @@
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("transpose_b", &transpose_b_));
+
+    OP_REQUIRES(ctx, fused_ops.size() == 1,
+                errors::InvalidArgument(
+                    "MklFusedMatMul must have only one argument: bias."));
+    OP_REQUIRES(
+        ctx, transpose_a_ == false,
+        errors::InvalidArgument("In[0] of MklMatMul can't be transposed."));
+  }
+
+  void Compute(OpKernelContext* ctx) override {
+    // FusedMatMul has 3 inputs: src, weights, bias
+    const Tensor& src_tensor = ctx->input(this->kInputIndexSrc);
+    const Tensor& weight_tensor = ctx->input(this->kInputIndexWeight);
+    const Tensor& bias_tensor = MklGetInput(ctx, this->kInputIndexBias);
+
+    MklDnnShape src_dnn_shape;
+    MklDnnShape weight_dnn_shape;
+    GetMklShape(ctx, this->kInputIndexSrc, &src_dnn_shape);
+    GetMklShape(ctx, this->kInputIndexWeight, &weight_dnn_shape);
+
+    // Get shapes of input tensors
+    auto src_tf_shape = src_dnn_shape.IsMklTensor() ? src_dnn_shape.GetTfShape()
+                                                    : src_tensor.shape();
+    auto weight_tf_shape = weight_dnn_shape.IsMklTensor()
+                               ? weight_dnn_shape.GetTfShape()
+                               : weight_tensor.shape();
+
+    // Check the constraint of input matrix and bias
+    OP_REQUIRES(ctx, TensorShapeUtils::IsMatrix(src_tf_shape),
+                errors::InvalidArgument("In[0] is not a matrix"));
+    OP_REQUIRES(ctx, TensorShapeUtils::IsMatrix(weight_tf_shape),
+                errors::InvalidArgument("In[1] is not a matrix"));
+    OP_REQUIRES(ctx, TensorShapeUtils::IsVector(bias_tensor.shape()),
+                errors::InvalidArgument("Biases must be 1D"));
+
+    // Expression: [batch, k] * [k, channel] + [channel] = [batch, channel]
+    //
+    // Get dimension size of each matrix, dim_pair[] is the location of k
+    // in the inputs, we have constraint that k of the two inputs are
+    // the same
+    const int dim_pair[] = {transpose_a_ ? 0 : 1, transpose_b_ ? 1 : 0};
+    const int batch = src_tf_shape.dim_size(1 - dim_pair[0]);
+    const int k = src_tf_shape.dim_size(dim_pair[0]);
+    const int channel = weight_tf_shape.dim_size(1 - dim_pair[1]);
+
+    OP_REQUIRES(ctx, k == weight_tf_shape.dim_size(dim_pair[1]),
+                errors::InvalidArgument("Matrix size-incompatible: In[0]: ",
+                                        src_tf_shape.DebugString(), ", In[1]: ",
+                                        weight_tf_shape.DebugString()));
+    OP_REQUIRES(ctx, bias_tensor.shape().dim_size(0) == channel,
+                errors::InvalidArgument(
+                    "Must provide as many biases as the channel size: ",
+                    bias_tensor.shape().DebugString(), " vs. ", channel));
+
+    // Create primitive for InnerProduct, it has such desc:
+    //   s[batch, k] * w[channel, k] + b[channel] = dst[batch, channel]
+    //    [n,     c] *  [oc,     ic] +  [x]       =    [n,     c]
```

Comment:
Should it be this way? (And should there be a `^T` at the end of the weight array the previous line?
```suggestion
    //    [n,    ic] *  [oc,     ic] +  [oc]      =    [n,    oc]
```

Reply:
Good suggestion, weight will be described as ^T format(but I only change the format desc to let MKL-DNN know it without really reorder) then I made the comment here. I will make some simplify.